### DISCUSSION TITLE
How to properly update plugins?

### DISCUSSION THREAD

**Discussion by klimmilk** [jhuckaby/Cronicle#722] [Q&A] [answered]
I can't figure out how to properly update plugins, preferably with zero downtime.
I have many scheduled tasks using my custom plugins. Quite often, changes need to be made to these plugins (sometimes only the logic may change, and sometimes the number of plugin parameters changes).
I'm currently using Cronicle in docker (but you can describe the solution for a non-docker installation) and haven't come up with anything yet other than:

1. Building new docker image
2. Creating backup from current container using `control.sh export`
3. Replacing old container with new one
4. Importing backup using `control.sh import`
5. Bulk update task parameters if needed via API

But this fill junky and this way I lose all task execution logs.

How to properly update plugins without losing your schedule and logs?

**Comment by jhuckaby** [comment]
Moving to discussion.  This is not an issue.

**Comment by jhuckaby** [comment, marked as answer]
Well, at my day job we just stop Cronicle entirely, update what we need, then start it again.  But I guess since you're using Docker that doesn't work, because if you stop Cronicle, the container exits?  I'm guessing that's what the problem is.

So, you'll have to script a sequence of actions using the REST API:

1. Disable the scheduler
2. Stop all jobs, OR wait for them to exit.
3. Update your Plugins.
4. Enable the scheduler again.

To enable/disable the scheduler, use the [update_master_state](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#update_master_state) API.

To wait for all jobs, poll the [get_active_jobs](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#get_active_jobs) API, and wait for it to return no results.  

To abort all jobs, first use the [get_active_jobs](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#get_active_jobs) API to get the IDs of all active jobs, then call [abort_job](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#abort_job) on each one to abort it.

Good luck!

**Reply by klimmilk** [reply]
Thanks for the answer!
Did I understand correctly that I can update plugins without stopping Сronicle completely? Just replace compiled files on the fly and re-enable the scheduler?

**Reply by jhuckaby** [reply]
Yes, my 4-step process shown above does not stop Cronicle.  It only pauses (disables) the scheduler via REST API call, then you can either wait for all active jobs to stop on their own, or abort them, then update your Plugin code, then enable the scheduler again.  This can all be done without stopping Cronicle, just using the REST API.
